kernel/userspace.c

```diff
@@ -94,6 +96,144 @@ struct perm_ctx {
 	struct k_thread *parent;
 };
 
+#ifdef CONFIG_DYNAMIC_OBJECTS
+#include <tommyhashlin.h>
+
+extern struct _k_object *_k_object_gperf_find(void *obj);
+extern void _k_object_gperf_wordlist_foreach(_wordlist_cb_func_t func,
+					     void *context);
+
+tommy_hashlin dyn_objects;
+int dyn_objects_init;
+
+static int init_dyn_objects(struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	tommy_hashlin_init(&dyn_objects);
+	dyn_objects_init = 1;
+	return 0;
+}
+
+SYS_INIT(init_dyn_objects, PRE_KERNEL_1, CONFIG_KERNEL_INIT_PRIORITY_OBJECTS);
+
+struct dyn_obj {
+	struct _k_object kobj;
+	tommy_node node;
+	u8_t data[];
+};
+
+static size_t obj_size_get(enum k_objects otype)
+{
+	switch (otype) {
+	case K_OBJ_ALERT:
+		return sizeof(struct k_alert);
+	case K_OBJ_MSGQ:
+		return sizeof(struct k_msgq);
+	case K_OBJ_MUTEX:
+		return sizeof(struct k_mutex);
+	case K_OBJ_PIPE:
+		return sizeof(struct k_pipe);
+	case K_OBJ_SEM:
+		return sizeof(struct k_sem);
+	case K_OBJ_STACK:
+		return sizeof(struct k_stack);
+	case K_OBJ_THREAD:
+		return sizeof(struct k_thread);
+	case K_OBJ_TIMER:
+		return sizeof(struct k_timer);
+	default:
+		return sizeof(struct device);
+	}
+}
+
+static int dyn_obj_cmp(const void *arg, const void *obj)
+{
+	struct _k_object *ko = (struct _k_object *)obj;
+
+	return arg != (void *)(ko->name);
+}
+
+struct _k_object *_k_object_find(void *obj)
+{
+	struct _k_object *ret;
+
+	ret = _k_object_gperf_find(obj);
+	if (ret) {
+		return ret;
+	}
+
+	if (!dyn_objects_init) {
+		return 0;
+	}
+
+	/* Not in the gperf table? Maybe it was a dynamic object */
+	return tommy_hashlin_search(&dyn_objects, dyn_obj_cmp,
+				    obj, tommy_inthash_u32((u32_t)obj));
+}
+
+void _k_object_wordlist_foreach(_wordlist_cb_func_t func, void *context)
+{
+	_k_object_gperf_wordlist_foreach(func, context);
+}
+
+
+void *k_object_alloc(enum k_objects otype)
+{
+	struct dyn_obj *obj;
+	int key;
+
+	/* Stacks are not supported, we don't yet have mem pool APIs
+	 * to request memory that is aligned
+	 */
+	__ASSERT(otype > K_OBJ_ANY && otype < K_OBJ_LAST &&
+		 otype != K_OBJ__THREAD_STACK_ELEMENT,
+		 "bad object type requested");
+
+	obj = k_malloc(sizeof(*obj) + obj_size_get(otype));
+	if (!obj) {
+		return 0;
```

Comment:
Should return NULL here. Maybe __ASSERT(0 && "could not allocate kernel object") as well?